### DISCUSSION TITLE
Can Flare crafting be upgraded and with many different NPCs with different skill sets and with ~'risky recipes'?

### DISCUSSION THREAD

**Discussion by WithinAmnesia** [flareteam/flare-game#859] [Q&A] [answered]
@dorkster Can Flare crafting be upgraded and many different N.P.C.s with different skill sets and with ~'risky recipes'? Does anyone around here have any ideas on how to improve / expand the crafting system to support hundred of recipes with ~thousands of possible ingredient combinations?  I am hoping to make Flare able to compete / be on par with crafting expectations of other grand scale R.P.G.s while not being too unrealistic to implement.

I got thinking in this comment for a similar vein of thought for how to improve the Flare crafting system: https://github.com/flareteam/flare-game/discussions/857#discussioncomment-1691336 + https://opengameart.org/forumtopic/can-flare-crafting-be-upgraded-and-with-many-different-npcs-with-different-skill-sets-and

Questions:
1. Can a list of available recipes for the specific crafting N.P.C. be within the same window as the crafting grid; a scroll-able N.P.C. available recipe list on the left-most side?
2. Can the crafting ingredients grid area be embiggened into something like a ~7x7 grid on the right-most side; up from the 4 square area?
3. Can an ~infinite scroll-able ingredients list in the middle of menu be made to work to vastly increase the crafting menu depth? 
4. Can X amount of Currency (I use Copper) work as a crafting ingredient along side ~regular trade-skill materials / ingredients?
5. Can there be different standardized sets of possible crafting recipe ~lists offered by different NPCs; e.g. a blacksmith, tailor, carver etc.? 
6. Can the crafting recipe output / reward also be a loot table / list system where there are arbitrary chances at different items on the loot list; to facilitate 'risky' recipes that act almost like Diablo II gambling where a chance system is used as a currency sink?*
--------------- 
*Also this 'risky recipe' system gives nigh-infinite replayability/versatility of enjoying content where every trade-skill relevant item can be potentially used to 'gamble' / craft upwards. Almost like in Path of Exile and Diablo II where items can be upgraded via crafting but at a ~large/great costs / lots of re-rolling from 'base' items.

**Comment by dorkster** [comment]
Flare's crafting system is more of a proof-of-concept for the MenuBook API than something designed to be scaled up like you describe. You should look at `books/crafting_alchemy.txt` and the scripts in `scripts/crafting/alchemy/` to get an idea of how much needs to be done for the current simple system.

The three objects you can have in books are text, images, and buttons. Considering each of these need to be placed individually, it's not ideal for large amounts of things. Although, it is possible to for buttons to open other books. So you can achieve the effect of multiple "pages" (see `books/wardens_notes.txt`).

1. Should be doable with the above described method of using multiple books. A button per recipe could be used to quickly craft things without having to place individual ingredients.
2. This is probably one of the bigger pain points. You would need define the image position and status flags for EACH ingredient in each of those 49 slots. The current crafting system is much smaller in scope, especially since the Empty Bottle is a required ingredient and can't go in any of the other slots. So we really only have 4 ingredients that can go into 3 slots.
3. Just like with 1, you might want separate pages for ingredients alone.
4. The actual crafting process is done with the Event system, so it's just a matter of requiring/removing some items and then rewarding the result. So in other words, yes, you can require currency for the crafting process in addition to ingredients.
5. There's no reason you can't use a different set of book(s) per NPC. Empyrean only has alchemy at the moment, but something like blacksmithing would get its own book.
6. As mentioned in 4, the Event system means you can use `reward_loot` for your crafting output.

**Reply by WithinAmnesia** [reply]
Thank you for your thoughts. I am trying to wrap my brain around how to make a proof of concept template for an expanded yet mechanically reasonable crafting system in Flare for even like ~16 different ingredients (~16 is the max I have witnessed in other games). I would like the 7x7 grid (49). Yet I feel like I need to be realistic as well to make this work. So perhaps it would be more reasonable to try for 4x4 grid (16 +1) with a top ingredient slot for currency (as how the empty bottom slot currently works) to start off with for version 1.0 of ~expanded Flare crafting?

I find that it tends to be hard to retroactively add scale. So I am hoping to go a bit larger (as what is within reason) so that the crafting system can support lots of possibilities. Also in having a large amount (even ~16+1) of crafting ingredients to start making a deep game with can pay off in the long run. As in with some cases it could make sense for say crafting to call for 10+ items. Such as with a hypothetical +1 to all resistances type of crafted item (e.g. utilizing every type / colour of gem *plus* a ~'base' item and or added crafting materials).

If a big variety of ingredient items can be supported in the crafting system yet be mechanically reasonable that could be quite successful without feature bloating. I am trying to figure out how to strike a balance of big possibilities for ingredients and outcomes and also still be realistic to develop / be efficient to work with. Hmmn, how can this work?

**Reply by WithinAmnesia** [reply]
Okay I have some more basic ideas. I am currently wondering on how this expanded crafting system might work but I'd like some thoughts on how this could be better:

Step one: Copy paste everything with the Alchemy N.P.C. and renamed it like 'Test Blacksmith' related; so new test N.P.C. and books and all that.
Step two: Hash out some rough placeholder blacksmith ingredients like 'Test Copper Bar', 'Test Grindstone', 'Test Leather Strip'.
Step three: Rename all of the Alchemy ingredients to the the test ingredients and change the 'crafted item' to a some kind of Test Copper Dagger or Test Copper Sword.
Step four: Break everything trying to make it bigger. More book pages with more ingredients slots and see how successful that is with ~16 ingredients slots and a page with recipes and required items and a page for each item recipe crafting grid? I am not sure how this will all look for I am kind of heading in blind.
Step five: Iterate, iterate, iterate until hopeful success and or acquisition of (hopefully) better ideas; then perhaps extrapolate those into a ~'version 2.0' / next version / next iteration of the idea(s).

Basic crafting goals:
1. ~16 ingredient slots (could even be a 4x4 cube / 4 rows of four, like current system +1 extra item slot wider with x 4 rows) and call that good enough for version 1.
2. The empty bottle slot could be a currency stack of ~1,000 currency (arbitrary amount).
3. Have 2 crafting recipe 'grid' pages. One for a dagger and one for a sword (sets up a template to expand later with more pages).
4. Have a starter page (or more pages with recipes lists if needed) with all recipes written down as notes possibly like (just to inform the player what is possible with this crafting N.P.C.).

------------------------------
"1. Test Dagger:
Materials Required:
- 1,000 Currency
- Test Copper Bar x 1
- Test Grindstone x1
- Test Leather Strip x1
------------------------------
2. Test Shortsword
Materials Required:
- 2,000 Currency
- Test Copper Bar x 2
- Test Grindstone x2
- Test Leather Strip x1"
------------------------------


It might work like this?:
Talk to 'Test Blacksmith'
*dialog splash of face image and usual dialog options*
-Blacksmithing
-Trade
-Talk (placeholder like ~Hello.)

*Select Blacksmithing*
Recipe List Page with all the possible recipes (or more pages after it if it runs out of space)
*Turn the page*
Looks like Alchemy but has 4x4 grid below a slot for ~X amount of currency (same item spot / slot as the empty bottle in the current Alchemy crafting example).

I am not entirely sure if we need 1 crafting grid page per 1 recipe combination and then add in more unique recipes via having a bunch of pages with 1 crafting grid page per 1 recipe combination; yet I feel this idea could possibly work? This approach might simplify adding in lots of recipes, whereas each page only has 1 possible combination like 'Test Copper Dagger Blacksmithing Page' and or Test 'Copper Shortsword Blacksmithing Page'. So that each recipe takes 1 page and to make more recipes more pages are added and maybe this can avoid a possible nightmare of the tackling combination issues / make life simpler / increased efficiency of unique recipe production?

This approach of 1 craft-able item combination per 1 crafting grid page is kind of like a visualized way of how dialog trees work where an N.P.C. (in other games like Baldur's Gate) has like a dialog tree and each item has an dialog question like 'what do you want the blacksmith to make?' *then a list of possible player dialogue choices appear such as* ~'please make dagger. or ~'please make sword' etc.; stuff like that. If 1 craft-able item combination per 1 crafting grid page approach is used then each item recipe would be its own little contained piece that can be copied and changed to make another unique recipe and the player can pick through all the unique recipes in list (or in Flare's case) individual pages of a ~book?

This approach of 1 craft-able item combination per 1 crafting grid page might end up with upwards of like ~100+ pages per trades-person N.P.C. but then it kind of feels like a 'book' and puts the sorting / recipe finding up to the player and potentially avoids the crafting combination cluster-fest crazy complexity of having all items be crafted on the same page XD. If we end up with like ~100+ pages the recipes could be like a cook book and have the recipe page(s) act as an index and have the recipes sorted alphabetically? Then possibly for an added ease of life option there could be a '~turn 10 pages' option? This could be kind of like how forums have the 'next page button' and like the 'big arrow button' to move a lot of pages at once? I am not sure how this would all work together but these are my current ~best ideas XP.

Does this make any sense? Is this reasonable or am I missing a few things along the way? I think you @dorkster made crafting in Flare to begin with? So if any one is an expert on what makes sense here in how to make crafting bigger it would be you. I kind of feel like I am taking wild guesses and stumbling around in the dark in a few areas and hoping for the best here XD.

**Reply by WithinAmnesia** [reply]
Update: "Should be doable with the above described method of using multiple books. A button per recipe could be used to quickly craft things without having to place individual ingredients." @dorkster 

This approach seems more reasonable in achieving a deep yet simple crafting system in Flare via using a "button per recipe could be used to quickly craft things without having to place individual ingredients". This 'craft recipe button' from 'books' approach looks like it can make everything less needlessly complex while also being more efficient to add / develop more / unique crafting recipes. The 'recipe list page' could also just act as a crafting page without grid squares and other complicated additions.

I am not sure if the button can be greyed out (like the skill / talent tree buttons) if the player does not have enough ingredients to craft the item. Although if the 'craft recipe button' could be greyed out (if the player lacks the required materials for said crafting recipe) that would be ideal; but if not it is not a game breaker. Even with just a simple message and or prompt informing that the player does not have enough materials to craft said recipe would be a success.

How can this system of using a "button per recipe could be used to quickly craft things without having to place individual ingredients" be set up successfully in Flare? Even just a template of two recipes / pages / crafting recipe 'book' feature demo or something to point myself in the right direction would be very helpful. I am just hoping to get the players able to collect ingredients and craft items from their adventures and I think this could be successful. Although I just have to figure out how to make it work now XD.

**Comment by WithinAmnesia** [comment, marked as answer]
Successful Crafting Approach: (It works now!) Via: https://github.com/flareteam/flare-engine/issues/1796#issuecomment-979806515

# Craft Dagger Test
[button]
button_pos=248,280
text=Forge Copper Off-Hand Dagger.
requires_currency=100
requires_item=3097:3
script=scripts/AGABAL/Crafting/Blacksmithing/Crafting_Test_Copper_Off-Hand_Dagger.txt

[event]
remove_currency=100
remove_item=3097:3
reward_item=3071
